enforcer/man/ods-enforcer-db-setup.8.in
.TH "ods-enforcer-db-setup" "8" "July 2016" "OpenDNSSEC" "OpenDNSSEC"
.SH "NAME"
.LP
.B ods\-enforcer\-db\-setup
\- OpenDNSSEC enforcer daemon initialization utility
.SH "SYNOPSIS"
.LP
.B ods\-enforcer\-db\-setup
.RB [ \-h ] 
.RB [ \-f ] 
.RB [ \-V ] 
.P
.SH "DESCRIPTION"
.LP
The ods\-enforcer\-db\-setup initializes the database used by the
ods\-enforcerd.  It should be used with care since all existing information
in the database schema will be deleted.  If you have existing zones stored
in this database this means that you loose the information on which zones
are signed with which keys.  Key material itself and signer information
are not removed.
.P
ods\-enforcer\-db\-setup is part of the OpenDNSSEC software.  For more
information, go to
.B http://www.opendnssec.org
and visit the Documentation page.
.P
.SH "OPTIONS"
.LP
.TP
.B \-h
Show this help.
.TP
.B \-f
Do not ask for confirmation.
.TP
.B \-V
Show version and exit.
.P
.SH "DIAGNOSTICS"
.LP
May log problems via standard syslog(8).
.SH "SEE ALSO"
.LP
ods\-control(8), ods\-enforcer(8), ods\-signerd(8),
ods\-signer(8), ods\-ksmutil(1), ods\-kasp(5),
ods\-kaspcheck(1), ods\-timing(5), ods\-hsmspeed(1),
ods\-hsmutil(1), opendnssec(7),
.B http://www.opendnssec.org/
.SH "AUTHORS"
.LP
.B ods\-enforcerd
was written by NLnet Labs as part of the OpenDNSSEC project.


NEWS
UNRELEASED:
* If listening port for signer is not set in conf file, the default value "15354" is used.
* Enforce and signconf tasks are now scheduled individually per zone. Resign
  per policy.
* zone delete removes tasks associated with zone from queue.
* Show help for ods-enforcer-db-setup
* In the kasp file, KSK/ZSK section, the algorithm length must be set now.
* The <Interval> tag had been deprecrecated, and is now no longer allowed to
  be specified.

OpenDNSSEC 2.0.1 - 2016-07-21

* Fixed crash and linking issue in ods-migrate.
* Fixed case where 2.0.0 could not read backup files from 1.4.10.
* Fixed bug in migration script where key state wasn't transformed properly.

OpenDNSSEC 2.0.0-1

* include db creation scripts in dist tarball needed for migration from 1.4.

OpenDNSSEC 2.0.0 - 2016-07-07

* OpenDNSSEC-99: Skip "are you sure" messages. Add --force and -f flag to
  ods-enforcer-db-setup and hsmutil purge
* OPENDNSSEC-808: Crash on query with empty query section (thanks Håvard Eidnes)
* OpenDNSSEC-771: Signer. Do not log warning on deleting a missing 
  NSEC3PARAM RR.
* OPENDNSSEC-801: Set AA flag on outgoing AXFR.
* SUPPORT-191: Regression, Must accept notify without SOA (thanks Christos Trochalakis)

OpenDNSSEC 2.0b1 - 2016-04-14

First public release of OpenDNSSEC.  Initial pre-releases have been
made to a smaller audience, this pre-release is explicitly made available
to all.  At this moment, there are no known functional bugs.  There are
naturally issues, especially to make working with OpenDNSSEC easier, however
none should prevent you to use OpenDNSSEC in production for the average
case, even though this is a pre-release.  Which is because of the still
limited documentation, and is not being run in production yet.

* The enforcer can no longer be run on a single policy at a time
  anymore.  An enforce run will always process all zones.
* The key generate method is at this time not available.
* The key export method will not allow you to export keys for all zones
  at once (--all flag) or for a particular type of key (--keystate).
  It will not export ZSK keys.
* The zonelist.xml in etc/opendnssec is no longer updated automatically,
  and by default works as if the --no-xml flag was specified.  Use
  --xml to the zone add command to update the zonelist.xml.  If updating
  the zonelist fails, the zone will still be added and not updated in
  the xml with future zone adds.
* Plugins directory renamed to contrib.
* Default signer working directory renamed from tmp to signer.
* Configure option --with-database-backend renamed --with-enforcer-database
* Zones on a manual rollover policy will not get a key assigned to them
  immediately.

OpenDNSSEC 2.0.0a5

Project transfer to NLnetLabs, performing code drop as-is for evaluation
purposes only.


OpenDNSSEC 2.0.0a4 (EnforcerNG branch) 

* SUPPORT-72: Improve logging when failed to increment serial in case 
  of key rollover and serial value "keep" [OPENDNSSEC-461].
* SUPPORT-114: libhsm: Optimize storage in HSM by deleting the public 
  key directly if SkipPublicKey is used [OPENDNSSEC-573].
* OPENDNSSEC-106: Add 'ods-enforcerd -p <policy>' option. This prompts the
  enforcer to run once and only process the specified policy and associated
  zones.
* OPENDNSSEC-330: NSEC3PARAM TTL can now be optionally configured in kasp.xml. 
  Default value remains PT0S.
* OPENDNSSEC-390: ods-ksmutil: Add an option to the 'ods-ksmutil key ds-seen'
  command so the user can choose not to notify the enforcer.
* OPENDNSSEC-430: ods-ksmutil: Improve 'zone add' - Zone add command 
  could warn if a specified zone file or adapter file does not exits.
* OPENDNSSEC-431: ods-ksmutil: Improve 'zone add' - Support default <input>
  and <output> values for DNS adapters.
* OPENDNSSEC-454: ods-ksmutil: Add option for 'ods-ksmutil key import' to
  check if there is a matching key in the repository before import.
* OPENDNSSEC-281: Enforcer NG: Commandhandler sometimes unresponsive.
* OPENDNSSEC-276, Enforcer NG: HSM initialized after fork().
* OPENDNSSEC-330: Signer Engine: NSEC3PARAM TTL is default TTL again, to
  prevent bad caching effects on resolvers.
* OPENDNSSEC-428: Add option for 'ods-ksmutil key generate' to take 
  number of zones as a parameter
* OPENDNSSEC-515: Signer Engine: Don't replace tabs in RR with whitespace.

Bugfixes:
* OPENDNSSEC-435: Signer Engine: Fix a serious memory leak in signature
  cleanup.
* OPENDNSSEC-463: Signer Engine: Duration PT0S is now printed correctly.
* OPENDNSSEC-466: Signer Engine: Created bad TSIG signature when falling back
  to AXFR.
* OPENDNSSEC-467: Signer Engine: After ods-signer clear, signer should not use
  inbound serial.


OpenDNSSEC 2.0.0a3 (EnforcerNG branch) - 2012-06-18
Bugfixes:
* SUPPORT-66: Signer Engine: Fix file descriptor leak in case of TCP write
  error [OPENDNSSEC-427].
* SUPPORT-71: Signer Engine: Fix double free crash in case of HSM connection
  error during signing [OPENDNSSEC-444].
* OPENDNSSEC-401: 'ods-signer sign <zone> --serial <nr>' command produces seg
  fault when run directly on command line (i.e. not via interactive mode)
* OPENDNSSEC-440: 'ods-ksmutil key generate' and the enforcer can create
  too many keys if there are keys already available and the KSK and ZSK use
  same algorithm and length
* OPENDNSSEC-424: Signer Engine: Respond to SOA queries from file instead
  of memory. Makes response non-blocking.
* OPENDNSSEC-425 Change "hsmutil list" output so that the table header goes
  to stdout not stderr
* OPENDNSSEC-438: 'ods-ksmutil key generate' and the enforcer can create
  too many keys for <SharedKeys/> policies when KSK and ZSK use same
  algorithm and length
* OPENDNSSEC-443: ods-ksmutil: Clean up of hsm connection handling
* Signer Engine: Improved Inbound XFR checking.
* Signer Engine: Fix double free corruption in case of adding zone with
  DNS Outbound Adapters and NotifyCommand enabled.

* Enforcer: Limit number of pregenerated keys when using <SharedKeys>.

enforcer/src/ods-enforcer-db-setup.c
/*
 * Copyright (c) 2014 Jerry Lundström <lundstrom.jerry@gmail.com>
 * Copyright (c) 2014 .SE (The Internet Infrastructure Foundation).
 * Copyright (c) 2014 OpenDNSSEC AB (svb)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

/**
 * OpenDNSSEC enforcer database setup tool.
 */

#include "config.h"

#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>

#include "daemon/engine.h"
#include "log.h"

#if defined(ENFORCER_DATABASE_SQLITE3)
#include <sqlite3.h>
#include "db/db_schema_sqlite.h"
#include "db/db_data_sqlite.h"
static const char** create = db_schema_sqlite_create;
static const char** drop = db_schema_sqlite_drop;
static const char** data = db_data_sqlite;
static sqlite3* db = NULL;
#elif defined(ENFORCER_DATABASE_MYSQL)
#include <mysql/mysql.h>
#include "db/db_schema_mysql.h"
#include "db/db_data_mysql.h"
static const char** create = db_schema_mysql_create;
static const char** drop = db_schema_mysql_drop;
static const char** data = db_data_mysql;
static MYSQL* db = NULL;
#endif

#define AUTHOR_NAME "Jerry Lundström"
#define COPYRIGHT_STR "Copyright (c) 2014 .SE (The Internet Infrastructure Foundation) OpenDNSSEC"

static void usage(FILE* out) {
    fprintf(out,
        "\nBSD licensed, see LICENSE in source package for details.\n"
        "Version %s. Report bugs to <%s>.\n",
        PACKAGE_VERSION,
        PACKAGE_BUGREPORT);
    fprintf(out, "--help,     -h: Print usage.\n");
    fprintf(out, "--version,  -V: Print version.\n");
    fprintf(out, "--force,    -f: Yes to all questions.\n");
}

static void version(FILE* out) {
    fprintf(out,
        "Database setup tool for %s version %s\n"
        "Written by %s.\n\n"
        "%s.  This is free software.\n"
        "See source files for more license information\n",
        PACKAGE_NAME,
        PACKAGE_VERSION,
        AUTHOR_NAME,
        COPYRIGHT_STR);
    exit(0);
}

static int connect_db(engineconfig_type* cfg) {
#if defined(ENFORCER_DATABASE_SQLITE3)
    if (!cfg->datastore) {
        return -1;
    }
    if (db) {
        return -1;
    }

    if (sqlite3_initialize() != SQLITE_OK) {
        return -1;
    }

    if (sqlite3_open_v2(cfg->datastore, &db,
        SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX | SQLITE_OPEN_CREATE,
        NULL) != SQLITE_OK)
    {
        return -1;
    }

    return 0;
#elif defined(ENFORCER_DATABASE_MYSQL)
    if (!cfg->datastore) {
        return -1;
    }
    if (db) {
        return -1;
    }

    if (mysql_library_init(0, NULL, NULL)) {
        return -1;
    }

    if (!(db = mysql_init(NULL))
        || !mysql_real_connect(db,
            cfg->db_host,
            cfg->db_username,
            cfg->db_password,
            cfg->datastore,
            cfg->db_port,
            NULL,
            0))
    {
        if (db) {
            mysql_close(db);
            db = NULL;
        }
        return -1;
    }

    return 0;
#else
    return -1;
#endif
}

static int disconnect_db() {
#if defined(ENFORCER_DATABASE_SQLITE3)
    if (db) {
        sqlite3_close(db);
        db = NULL;
    }

    sqlite3_shutdown();

    return 0;
#elif defined(ENFORCER_DATABASE_MYSQL)
    if (db) {
        mysql_close(db);
        db = NULL;
    }

    mysql_library_end();

    return 0;
#else
    return -1;
#endif
}

static int db_do(const char *sql, size_t size) {
#if defined(ENFORCER_DATABASE_SQLITE3)
    sqlite3_stmt* stmt = NULL;

    if (!db) {
        return -1;
    }
    if (!sql) {
        return -1;
    }

    if (sqlite3_prepare_v2(db, sql, size, &stmt, NULL) != SQLITE_OK
        || sqlite3_step(stmt) != SQLITE_DONE)
    {
        if (stmt) {
            sqlite3_finalize(stmt);
        }
        return -1;
    }
    sqlite3_finalize(stmt);

    return 0;
#elif defined(ENFORCER_DATABASE_MYSQL)
    if (!db) {
        return -1;
    }
    if (!sql) {
        return -1;
    }

    if (mysql_real_query(db, sql, size)) {
        return -1;
    }

    return 0;
#else
    return -1;
#endif
}

static int db_do2(const char** strs) {
    char sql[4096];
    char *sqlp;
    int ret, left, i;

    for (i = 0; strs[i]; i++) {
        left = sizeof(sql);
        sqlp = sql;

        for (; strs[i]; i++) {
            if ((ret = snprintf(sqlp, left, "%s", strs[i])) >= left) {
                return -1;
            }
            sqlp += ret;
            left -= ret;
        }

        if (db_do(sql, sizeof(sql) - left)) {
            return -1;
        }
    }

    return 0;
}

int main(int argc, char* argv[]) {
    int c, options_index = 0;
    static struct option long_options[] = {
        {"help", no_argument, 0, 'h'},
        {"version", no_argument, 0, 'V'},
        {"force", no_argument, 0, 'f'},
        { 0, 0, 0, 0}
    };
    int user_certain;
    int force = 0;
    engineconfig_type* cfg;
    const char* cfgfile = ODS_SE_CFGFILE;

    ods_log_init("ods-enforcerd", 0, NULL, 0);

    while ((c=getopt_long(argc, argv, "hVf",
        long_options, &options_index)) != -1) {
        switch (c) {
            case 'h':
                usage(stdout);
                exit(0);
            case 'V':
                version(stdout);
                exit(0);
	    case 'f':
		force = 1;
		break;
            default:
                exit(100);
        }
    }

    if (!force) {
        printf("*WARNING* This will erase all data in the database; "
	       "are you sure? [y/N] ");
        user_certain = getchar();
        if (user_certain != 'y' && user_certain != 'Y') {
            printf("Okay, quitting...\n");
            return 0;
        }
    }

    cfg = engine_config(cfgfile, 0, NULL);
    if (engine_config_check(cfg) != ODS_STATUS_OK) {
        engine_config_cleanup(cfg);
        fprintf(stderr, "Error: unable to load configuration!\n");
        return 1;
    }

    if (connect_db(cfg)) {
        engine_config_cleanup(cfg);
        fprintf(stderr, "Error: unable to connect to database!\n");
        return 2;
    }

    /*
     * Drop existing schema.
     */
    if (db_do2(drop)) {
        fprintf(stderr, "Error: unable to drop existing schema!\n");
        disconnect_db();
        return 3;
    }

    /*
     * Create new schema.
     */
    if (db_do2(create)) {
        fprintf(stderr, "Error: unable to create schema!\n");
        disconnect_db();
        return 4;
    }

    /*
     * Insert initial data.
     */
    if (db_do2(data)) {
        fprintf(stderr, "Error: unable to insert initial data!\n");
        disconnect_db();
        return 5;
    }

    engine_config_cleanup(cfg);
    printf("Database setup successfully.\n");
    return 0;
}
